### ISSUE TITLE
Observing high volume of Decrypt messages from Sops-operator In Kubernetes EKS 1.25 

### ISSUE THREAD

**Issue by sakomws** [isindir/sops-secrets-operator#137]
Hi!

We are observing high volume of Decrypt messages sending from Sops operator to KMS In Kubernetes EKS 1.25 which shows up in Sops operator debug logs as below:
```
{"level":"info","ts":"2023-06-21T18:21:03Z","logger":"controllers.SopsSecret","msg":"Secret successfully refreshed","secret":"x","namespace":"x"}
```

Looking in the code, what triggers it the difference in secret objects:
https://github.com/isindir/sops-secrets-operator/blob/0992896af64ac806725292db269b39ebc8421026/internal/controllers/sopssecret_controller.go#L172

We are in latest release of Sops-operator v0.9.0 as well. 

Looking for any guidance or config changes would help, thank you!

**Comment by isindir** [comment]
@sakomws It is a bug in controller, I'll take a look.

**Comment by sakomws** [comment]
> @sakomws It is a bug in controller, I'll take a look.

Appreciate getting back to us @isindir ! Do you have approximate timeline for a fixed? Any workaround you would recommend in the meantime?

**Comment by isindir** [comment]
@sakomws I'll try to address it these weekends, but can not guarantee.

**Comment by sakomws** [comment]
thank you so much, appreciate for amazing project!

**Comment by isindir** [comment]
@sakomws do you have comparable examples of the output from controller running in EKS 1.24 and EKS 1.25 ? How much more output the controller produces in 1.25 ? The quick fix I thought to implement did not fly very well - it broke functionality, I'm thinking to tune logging itself, but it will take longer to implement.

**Comment by isindir** [comment]
@sakomws I have released new version of the controller with reduced amount of logging by default, please upgrade and check.  Please see flags available for configuring chart in logging section of `values.yaml`.

**Comment by sakomws** [comment]
Thats awesome! Going to try out tomorrow and let you know! Appreciate your help!

**Comment by isindir** [comment]
@sakomws how was your testing going on? Did my fix help you nad/or what was the issue ?

**Comment by isindir** [comment]
Closing issue

**Comment by signaleleven** [comment]
Not sure if it's related to this. I upgraded my clusters from k8s 1.24 to 1.26 (through 1.25) and since then every version of the controller (I tried the helm chart from 0.10.x to 0.15.x) is infinitely looping through all the secrets, to reconcile them, even if they did not change. 
I noticed this through a spike in KMS requests in AWS (so, even a cost impact!).
I cannot identify what difference does the the controller see in the existing secrets, and switching the logging to debug does not help.

```
﻿﻿2023-08-14T12:23:38.958626881Z stderr F 1.6920158189585416e+09	INFO	controllers.SopsSecret	Processing	{"sopssecret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:38.958916995Z stderr F 1.692015818958837e+09	INFO	controllers.SopsSecret	Secret already exists and needs to be refreshed	{"secret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:38.968911859Z stderr F 1.6920158189687552e+09	INFO	controllers.SopsSecret	Secret successfully refreshed	{"secret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:38.984599449Z stderr F 1.6920158189844537e+09	INFO	controllers.SopsSecret	SopsSecret is Healthy	{"sopssecret": "test-provider-tasker/test-provider-tasker-secrets"}
2023-08-14T12:23:38.987191391Z stderr F 1.6920158189870803e+09	INFO	controllers.SopsSecret	Reconciling	{"sopssecret": "odt-api/odt-api-secrets"}
2023-08-14T12:23:39.017766001Z stderr F 1.692015819017612e+09	INFO	controllers.SopsSecret	Entering template data loop	{"sopssecret": "odt-api/odt-api-secrets"}
2023-08-14T12:23:39.017933329Z stderr F 1.6920158190178583e+09	INFO	controllers.SopsSecret	Processing	{"sopssecret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:39.018186599Z stderr F 1.692015819018117e+09	INFO	controllers.SopsSecret	Secret already exists and needs to be refreshed	{"secret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:39.025256507Z stderr F 1.6920158190251074e+09	INFO	controllers.SopsSecret	Secret successfully refreshed	{"secret": "REDACTED", "namespace": "REDACTED"}
2023-08-14T12:23:39.043048893Z stderr F 1.6920158190428874e+09	INFO	controllers.SopsSecret	SopsSecret is Healthy	{"sopssecret": "odt-api/odt-api-secrets"}
2023-08-14T12:23:39.043420047Z stderr F 1.692015819043324e+09	INFO	controllers.SopsSecret	Reconciling	{"sopssecret": "user-config-api/user-config-api-secrets"}

```

**Comment by isindir** [comment]
@signaleleven - can you please scale controller down to 0 replicas and check if one of the target secrets and one of the sops-secrets is not changing after that? By change - I mean any change to metadata or status.

**Comment by signaleleven** [comment]
Hi! 
I upgraded one cluster to 1.27 just to see if that helped: it doesn't.
Answering your questions

I did scale the deployment to 0 (if anything, to avoid the AWS costs!) and queried a secret that is managed by the controller for changes. It does not changes.

I do use argocd to deploy the sopssecret objects, but they are not being updated (the argocd applications are not set to autoheal, in case of changes I would see the objects being out of sync, and I do not)

The secrets are all one-to-one with their sealed counterparts.

The values.yaml for the controller is absolutely empty (so default values) save for the annotations for the service account (for EKS, so it gets assigned an IAM role).
I changed the logging setting to raise the logging to debug, in order to maybe see what differences does the controller see. I saw some debug messages but nothing suggesting what is the difference it sees.
I rolled back that change, so I confirm my only deployment value is the service account annotation.

**Comment by erikzele** [comment]
> I do use argocd to deploy the sopssecret objects, but they are not being updated (the argocd applications are not set to autoheal, in case of changes I would see the objects being out of sync, and I do not)

I suspect there is a constant competition between 2 or more tools updating same `sopssecret` or child `secret`, and by that I mean any field including metadata or status, where it can be timestamp or any other field.

Can you please observe that by watching (using `-w` flag) sopssecret that it's content in the cluster is not updated.

Can you please share what settings did you use to see debug messages ?

**Comment by isindir** [comment]
> I do use argocd to deploy the sopssecret objects, but they are not being updated (the argocd applications are not set to autoheal, in case of changes I would see the objects being out of sync, and I do not)

I suspect there is a constant competition between 2 or more tools updating same `sopssecret` or child `secret`, and by that I mean any field including metadata or status, where it can be timestamp or any other field.

Can you please observe that by watching (using `-w` flag) sopssecret that it's content in the cluster is not updated.

Can you please share what settings did you use to see debug messages ?

**Comment by signaleleven** [comment]
So, I set up two terminals with `kubectl get -w -A [sopssecrets|secrets]` and scaled up sops-secrets-controller

The sopssecrets started in "Child secret update error", when I start the controller they are constantly updated in state "Healty". A never ending stream. The content, AFAICT, is unchanged. The decoded secrets do not change.
As I said, argocd is present on the cluster, and from the UI I don't see any change.

However... 

<img width="1446" alt="image" src="https://github.com/isindir/sops-secrets-operator/assets/1616975/1821a54d-10b7-47f5-92ca-2b2d5bf203a2">

ArgoCD is indeed reacting to the "changes" triggered by sops-secrets-controller (the timing of the peak in logs match the time the controller was active)

I don't really see what it is reacting too, nor I can explain why this was not a problem until the upgrade to k8s 1.25.

**Comment by erikzele** [comment]
I think if you create sopssecret outside of the argo, that should not be reconciling that often.

**Comment by signaleleven** [comment]
yeah, but that's not an option, 99% of what runs in the cluster is managed by argocd, and it was fine in the past. I will try to understand if and why argocd is reacting to what change and possibly exclude the field from reconciliation with an annotation. The problem is comparing a sopssecret object before and after seems to show that there are no changes (other than the status at the beginning,going to "Child secret update error" to "Healthy" and back when I stop the controller.

**Comment by signaleleven** [comment]
I also tried to observer what happens to the children, the decrypted secrets. I deleted the `frontend-secrets-test` while the controller was scaled to zero and I started watching for secrets

```
$▶ kubectl get -w secrets -n frontend
NAME                        TYPE                DATA   AGE
frontend-secrets             Opaque              3      2d
# This was the only secret. The output did not continue, until I scaled up the controller
frontend-secrets-test        Opaque              3      0s
frontend-secrets-test        Opaque              0      0s
frontend-secrets-test        Opaque              3      3s
```

When I scaled up the controller, I had those 3 lines. I don't know why the secret got temporarily emptied with zero data, but after those 3 seconds nothing got updated anymore (whereas the two corrisponding sopssecrets behaved like in the previous comment).

The secret itself has nothing remarkable that might trigger the controller to think it needs a reconciliation

```
$▶ kubectl get -o yaml secret frontend-secrets-test -n frontend
apiVersion: v1
data:
  REDACTED: REDACTED 
  REDACTED: REDACTED
  REDACTED: REDACTED
kind: Secret
metadata:
  creationTimestamp: "2023-08-16T13:03:56Z"
  name: frontend-secrets-test
  namespace: frontend
  ownerReferences:
  - apiVersion: isindir.github.com/v1alpha3
    blockOwnerDeletion: true
    controller: true
    kind: SopsSecret
    name: frontend-secrets-test
    uid: dea69596-0740-4657-a9fd-a282c31c7549
  resourceVersion: "52867091"
  uid: f9b3cdeb-1079-46e1-915a-ce1119381580
type: Opaque
```

**Comment by pauldn-wttj** [comment]
I'm having the same issues of infinite loop after upgrading the operator on EKS 1.25 with ArgoCD, very similar to what @signaleleven describes.

- I was on EKS 1.23, upgrading to 1.25 (through 1.24).
- `sops-secrets-operator` was deployed using chart version 0.10.8 so I upgraded to 0.12.5 to match Kubernetes version.
- After that, the operator was logging a lot of `Secret already exists and needs to be refreshed`, refreshing secret until healthy and it appeared it was looping indefinitely. I can see that `SopsSecret` resources are constantly updated with their metadata.resourceVersion` updating very frequently.
- I tried rolling back to Helm chart 0.10.8 but it did not change.
- If I let the `sops-secrets-operator` pod running, `SopsSecret` usually have `message: Healthy` in their status. However, once I scale it down to 0 pods, I am seeing `Child secret update error` too.

An additional detail, I have another EKS cluster that I recently upgraded from 1.23 to 1.25, on which the `sops-secret-operator` has not yet been upgraded (so it is still using Helm chart 0.10.8) and I am not seeing the issue: no loop, no update of the `SopsSecret`.

I'll try looking into what might cause the infinite loop.

**Comment by signaleleven** [comment]
It looks @pauldn-wttj and I are exactly in the same situation. I'd like to bring up a new cluster (without argocd, without anything other than the controller) to see how it behaves there. I'll report if I manage to do it.

**Comment by signaleleven** [comment]
Ok, on a brand new EKS cluster (1.27 + sops-secrets-controller helm 1.15.4) the problem is *not* present.

**Comment by signaleleven** [comment]
Even with argocd (not managing any of the secrets) there is no loop on the new cluster. Scaling the controller to 0, and then to 1 makes exactly one refresh of each secret (but they never go in "child secret update error")

The only difference between this test cluster and those where I observe the problem is that this started at 1.27, the others went through an upgrade.

**Comment by pauldn-wttj** [comment]
There is probably a component on my cluster which is triggering the operator to try to reconcile `SopsSecrets` unnecessarily, but after looking at the code, I observed things that might be partly causing the issue, or maybe I am missing some things.

For instance, I have a `SopsSecret` as following:

```yaml
apiVersion: isindir.github.com/v1alpha3
kind: SopsSecret
metadata:
    name: secret-env
spec:
    suspend: false
    secretTemplates:
        - name: secret-env
          stringData:
            PASSWORD: ENC[AES256_GCM,data:...,type:str]
sops:
    kms: <KMS configuration>
    ...
    encrypted_suffix: stringData
    version: 3.7.3
```

This creates the following `Secret`:

```yaml
apiVersion: v1
data:
  PASSWORD: ...
kind: Secret
metadata:
  name: secret-env
  namespace: <namespace>
  ownerReferences:
  - apiVersion: isindir.github.com/v1alpha3
    blockOwnerDeletion: true
    controller: true
    kind: SopsSecret
    name: secret-env
    uid: <sopssecret-uid>
  resourceVersion: "<resource-version>"
  uid: <secret-uid>
type: Opaque
```

The problem I observed was that [refreshKubeSecretIfNeeded](https://github.com/isindir/sops-secrets-operator/blob/321cb0e4f173f8bcc13a5399e87a880f97d198ff/controllers/sopssecret_controller.go#L94) was always updating the `SopsSecret` even though nothing changes. The operator always notices a change between the current `Secret` in the cluster and the `Secret` it would create from the `SopsSecret` template, and this condition is always `true`: https://github.com/isindir/sops-secrets-operator/blob/321cb0e4f173f8bcc13a5399e87a880f97d198ff/controllers/sopssecret_controller.go#L162

due to those two lines:

https://github.com/isindir/sops-secrets-operator/blob/321cb0e4f173f8bcc13a5399e87a880f97d198ff/controllers/sopssecret_controller.go#L152-L153

To explain why I think that it's the case: 
- `copyOfKubeSecretInCluster` is supposed to be the `Secret` it would create from the template, it has `stringData` **and** no `data`
- `kubeSecretInCluster` is supposed to be the actual `Secret` in the cluster, it has `data` (since once applied, a Kubernetes `Secret` has `data`) **and** no `stringData`.

So the `DeepEqual` will never return `true`.

To sum up:
- There is probably some component on my cluster triggering the operator to try to reconcile
- The operator enters the main loop and updates the `SopsSecret` resource even though it shouldn't.
- It triggers the previous component to update the resource
- Repeat indefinitely

Again, I might be missing something, Go is not my area of expertise!

**Comment by pauldn-wttj** [comment]
For now we ended up using the patch mentioned by @maxd-wttj to ensure the operator would not try to reconcile unnecessarily due to a false positive when checking which resources need updates. This does not explain everything since we still don't know what in our cluster causes the operator to try to reconcile but it's avoiding us infinite loops.

If anyone has a better idea / understanding, feedback are welcome! :slightly_smiling_face:

**Comment by pauldn-wttj** [comment]
I agree with you, it does not explain why reconciliations are re-triggered after the operator updates secrets which is the root cause of the problem. However, with our limited time (and understanding!) we could not find where it could come from. It's all the more confusing that we do have two EKS clusters in the same versions, using a (different) custom KMS key for secret encryption, where operator is working fine on one, and looping on the other.

**Comment by erikzele** [comment]
I personally deploy it atm only on rpi k3s cluster with age key and never seen this behaviour, so would be difficult to reproduce in practice. Out of interest, are you using kms or any other encryption mechanism, like age or gpg?

**Comment by signaleleven** [comment]
I use KMS encryption (and deploy on EKS, as mentioned)

**Comment by signaleleven** [comment]
Update: the controller on my brand new 1.27 cluster started misbehaving several hours after provisioning. It was outside of working hours, nobody was touching it or deploying anything. 
I start to feel that this depends on something on the EKS control plane and we see different behaviours because different clusters might be running on slightly different versions of the backend (of course AWS only shows us the major version. 1.2x for the control plane).
I see no discernible pattern.

**Comment by erikzele** [comment]
May be its worth opening a case with AWS as well ? They might be able to correlate to some EKS related events potentially.. Also if secrets are not changed often, atm it could be possible as a hotfix to scale down controller to 0 after secreys are created, do not remember if I implemented such annotation or field to stop reconciling specific sops objects when needed, if that is implemented - might be quite usefull..

**Comment by erikzele** [comment]
ah, yes - one can just disable recociliation for semi-static sops objects, by setting `suspend` field value to `true`...

**Comment by signaleleven** [comment]
That's my strategy now, I keep the controlled scaled to 0, and if I need to deploy new secrets or upgrade existing ones I scaled it up temporarily.

I'm not entirely sure what to ask AWS support though... I could point to this issue, but I'm afraid it'd be dismissed as application caused.

Re: "suspended" - interesting, I'll look it up

**Comment by pauldn-wttj** [comment]
The problem still ended up occurring on our side even with the patch we tried to implement (#145).

Here are more tests we made:
- The application-controller from Argo seemed to reconcile a lot so we figured it might cause the sops-secrets-operator to reconcile, hence the infinite loop. We tried upgrading to ArgoCD 2.8 which provides a way to ignore resource updates on some resources (see [documentation](https://argo-cd.readthedocs.io/en/release-2.8/operator-manual/reconcile/#system-level-configuration)) but it had no effect although the argocd-application-controller did ignore all updates made by the sops-secrets-operator according to logs: 
```
{"api-version":"isindir.github.com/v1alpha3","kind":"SopsSecret","level":"debug","msg":"Ignoring change of object because none of the watched resource fields have changed","name":"app-env","namespace":"namespace","server":"https://kubernetes.default.svc","time":"2023-08-31T15:26:36Z"}
```
- We tried removing the final `Update` the controller does in the reconciliation here:
https://github.com/isindir/sops-secrets-operator/blob/2d011cecdc9658d6882ec35ab7c04c919bf8f7ab/internal/controllers/sopssecret_controller.go#L107-L108
and the infinite loop was gone so our main theory is that the final `Update` is causing some other components (unlikely to be ArgoCD) to reconcile and re-trigger the sops-secrets-operator reconciliation. We couldn't find anything in EKS logs unfortunately. :confused:

**Comment by isindir** [comment]
@pauldn-wttj interesting, I have tried following #146, but seems it is failing the test of managed secret to be refreshed if updated outside of the controller

**Comment by erikzele** [comment]
@pauldn-wttj I need to go through the logic again and see if reconcile loop fails on the first sub-secret which has broken config or has issues on update and stops there. If that is the case I hope to be able to set `Healthy`/`Failed` status on each sub-secret, remove final update and set controller to ignore status updates as in #146 .

**Comment by isindir** [comment]
I think the repeated reconciliation is more likely related to the design of the operator:
* I have only one reconciliation loop which reacts on changes for both `sops` and `secret` objects
* `sops` objects have both fields: `resourceVersion` and `generation`, where `secret` objects by API documentation have both in reality have just `resourceVersion`

I think both of these mean that I can not use predicates to filter changes to status of `sops`, as `secret` would stop triggering any changes due to the fact of missing `generation` on it, and one loop means that filter applies to both object types.

On the other hand, if there is change to `sops` object, it triggers the following events:
* updates for all the child `secrets`
* for each child `secret` update `sops` object is being reconciled
* final `sops` status change (error or success) is triggering an extra reconciliation of the `sops` object as well

So my theory is - on EKS it hits some sort of race condition which leads to infinite reconciliation loop.

There are several possible options here:
* split single reconciliation loop into multiple - `sops`, `secret` (severe re-design)
* try to move `Helthy` status update to the place before iterating all the Templates (so it will be re-triggered at different time of the logic - not sure if this will help, but easy to change and try)
* Make `Healthy` status update optional via extra (non-default flag) for the operator (hack to hotfix the issue)

@pauldn-wttj  is it possible for you to try 2nd option in eks ?

**Comment by maxd-wttj** [comment]
> @pauldn-wttj is it possible for you to try 2nd option in eks ?

Not @pauldn-wttj but we are coworkers working on this issue.
I'm just tried moving the Healthy status update before the template iteration (right there: https://github.com/isindir/sops-secrets-operator/blob/1068159d93c84f3fa6887e6f2c39915b8537f548/internal/controllers/sopssecret_controller.go#L77C1-L77C1)

Unfortunately the loop still occurs.

**Comment by maxd-wttj** [comment]
It might be a bit hacky, but conditionally updating the status instead of updating it everytime seems to do the trick:
```
if encryptedSopsSecret.Status.Message != "Healthy" {
    encryptedSopsSecret.Status.Message = "Healthy"
    _ = r.Status().Update(context.Background(), encryptedSopsSecret)
}
```
Apparently this does not break the tests.

**Comment by isindir** [comment]
@maxd-wttj I have added above code in #147, please try it and after a week or 2 please confirm if that fixed the issue and I can close it.

**Comment by pauldn-wttj** [comment]
Thanks @isindir, we're upgrading the operator this morning and will let you know!

**Comment by maxd-wttj** [comment]
Hello @isindir, it's been a week since we applied the patch, and we haven't encountered any infinite reconciliation loop yet!

**Comment by isindir** [comment]
@maxd-wttj, thank you for confirmation - additionally to the fix in #147 - I have done conditional change via newly added function to all places, where `status.message` was refreshed with the same value as already set, please see in #148.

**Comment by isindir** [comment]
closing case

**Comment by signaleleven** [comment]
Just for completeness, I confirm that the latest version works on all my clusters without issues! THANK YOU for your support @isindir
